### DISCUSSION TITLE
Global Keyboard Events and AddClassHandler

### DISCUSSION THREAD

**Discussion by LaurentInSeattle** [AvaloniaUI/Avalonia#17432] [Q&A] [answered]
Hi all, 

My app' is managing global keyboard events using AddClassHandler at the top level, like: 

`
InputElement.KeyDownEvent.AddClassHandler<TopLevel>(OnKeyDown, handledEventsToo: true);
`
That's works. 

Q1) Is there a better way to handle them, knowing they should be received even when non focused ?
 
Q2) It seems like there is no way to remove this hook when it becomes not needed any longer? 
I could not find any suitable remove, unsubscribe, unregister... API. 

Thanks!

**Comment by maxkatz6** [comment, marked as answer]
> Q1) Is there a better way to handle them, knowing they should be received even when non focused ?

If you own the window, it would be better to handle this event on the window directly with Tunnel strategy. 
`window.AddHandler(InputElement.KeyDownEvent, OnKeyDown, Tunnel, handledEventsToo: true)` (pseudocode).

This way, you can avoid static global event handlers, and still handle events first.

> knowing they should be received even when non focused ?

This will work for unfocused controls, but focused window. You can't intercept input outside of your app this way. There is no API for that.

> It seems like there is no way to remove this hook when it becomes not needed any longer?

It returns a disposable. You can dispose this subscription.

**Reply by LaurentInSeattle** [reply]
Thanks for the quick reply. 
Your suggested alternate solution does the job, avoiding statics... Good.

The doc' page about AddClassHandler at https://reference.avaloniaui.net/api/Avalonia.Interactivity/RoutedEvent/C7C528DD only mentions: Returns: IDisposable  with no description, comments, hints... nothing.
